feat: JWT cache implementation based on sieve algorithm

mkleczek opened the pull request:
Draft version of JWT cache implementation based on https://cachemon.github.io/SIEVE-website/blog/2023/12/17/sieve-is-simpler-than-lru/ algorithm

mkleczek commented:
This PR contains:

1. Refactoring and some cleanup of JWT handling code:
    - Moved JWT parsing and validation to a separate module Auth.JWT
    - Split JWT decoding, parsing and signature validation, claims validation into separate functions
    - Instead of caching `AuthResult` cache decoded claims (which signature was verified). Validating claims and determining role is done *after* cache lookup
    - Cleaned up API so that usage of it is simplified: `lookupJwtCache cache key >>= parseClaims configJwtAud time`
    - handling of JwtCacheState initialization and updates of configuration is encapsulated in Auth.JwtCache module
2. Generic scalable cache implementation based on _stm_,  _stm-hamt_ and _sieve_ algorithm. It also provides usage stats (ie. hit ratio, evictions count, size)

steve-chavez commented:
Very interesting! :eyes: :eyes: 

> It also provides usage stats (ie. hit ratio, evictions count, size)

Maybe we could expose those as [metrics](https://docs.postgrest.org/en/v13/references/observability.html#metrics). That would help with passing code coverage too, since it looks like it's detecting those functions as dead code:

> postgrest-13.1-inplace: src/PostgREST/Cache/Sieve.hs:124:1: delete
> postgrest-13.1-inplace: src/PostgREST/Cache/Sieve.hs:133:1: deleteIO
> postgrest-13.1-inplace: src/PostgREST/Cache/Sieve.hs:136:1: resetIO
> postgrest-13.1-inplace: src/PostgREST/Cache/Sieve.hs:142:1: accessStats
> postgrest-13.1-inplace: src/PostgREST/Cache/Sieve.hs:148:1: evictionsCount

I also see the JWT loadtest failing on CI https://github.com/PostgREST/postgrest/actions/runs/15030111955?pr=4084:

> Options "http://postgrest/authors_only": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

Not sure what's going on there, but you should be able to reproduce it locally with `postgrest-loadtest -k jwt`.

steve-chavez commented:
> postgrest-loadtest -k jwt gives me:
> delaying data to/from PostgREST by 0ms
> jwt": -c: line 1: unexpected EOF while looking for matching `"'

Weird, it's like you're using and old version of the nix tools and it's ignoring the `-k jwt` arg. On my machine I get:

```bash
$ postgrest-loadtest -k jwt
Created 50000 targets in ./test/load/gen_targets.http (0.79s)
...

## this is the ouput you're getting
$ postgrest-loadtest -k mixed # same as just "postgrest-loadtest"
delaying data to/from postgres by 0ms 
```

Maybe try going out of `nix-shell` and running it again.

mkleczek commented:
Yeah, that was it, thanks.

I have serious doubts about `postgrest-loadtest -k jwt`. When testing `main` on my machine (beefy MacBook) there is no difference in results between `PGRST_JWT_CACHE_MAX_LIFETIME=86400` and `PGRST_JWT_CACHE_MAX_LIFETIME=0`. I guess that's because symmetric crypto is used to generate/validate tokens. That's probably too fast to be worth caching and show any difference between cached and not cached access. I think meaningful tests have to be based on asymmetric cryptography.

Nevertheless there indeed seems to be an issue with implementation in this PR: all percentiles (ie. median up to 99th) show _better_ results than `main` but _max_ is 30s (which causes timeouts and skews average response time). Looking at it right now.

steve-chavez commented:
> That's probably too fast to be worth caching and show any difference between cached and not cached access. I think meaningful tests have to be based on asymmetric cryptography.

Agree, we can change that.

> When testing main on my machine (beefy MacBook) there is no difference in results between PGRST_JWT_CACHE_MAX_LIFETIME=86400 and PGRST_JWT_CACHE_MAX_LIFETIME=0.

Note that the jwt loadtest actually generates unique JWTs , it was mainly done to test the jwt decoding perf while also ensuring the JWT cache purging doesn't slow things down (https://github.com/PostgREST/postgrest/pull/4034).

The "mixed" loadtest does have hardcoded JWTs, and the perf looks more or less maintained (or slightly better).

mkleczek commented:
@steve-chavez @wolfgangwalther @taimoorzaeem

I am not sure what to do with outstanding test coverage issues. This is about `delete` and `deleteIO` not being used. And indeed - right now there is no explicit cache entry removal anywhere. These functions are meant to be used in case of claims validation errors.

The question is: how to handle invalid JWTs? There are several options:

1. Do not cache invalid JWTs at all
2. Do not cache JWTs that cannot be parsed or have invalid signature. Cache all other JWTs, also JWTs containing invalid claims.
3. Cache all token parsing and signature validation results - this is currently implemented option (ie. the type of values in the cache is `Either Error (JSON.Object)`
4. Cache all tokens that can be parsed (but possibly having invalid signature).

I've decided option three makes sense as it also speeds up repeated invalid token requests. OTOH - it opens up possibility of _cache filling_ attacks using randomly generated tokens. It is disputable if that is a bigger problem than overloading CPU with eg. JWTs signed with random key.

It looks to me option 4 would be the best compromise because it would prevent filling up cache with random garbage but would offload signature validation. I haven't implemented it as I don't know how to split parsing and signature validation with JOSE.

WDYT?

mkleczek commented:
> Very interesting! 👀 👀
> 
> > It also provides usage stats (ie. hit ratio, evictions count, size)
> 
> Maybe we could expose those as [metrics](https://docs.postgrest.org/en/v13/references/observability.html#metrics).

@steve-chavez 

Done.

mkleczek commented:
> But do we really need to? Parsing a JWT is really simple - but creating parseable tokens, with invalid signatures is just as easy. So I don't really see the additional value of option 4 vs option 3. If somebody wants to fill the cache, they can do so easily.

Indeed - any random bytes are a signature that consumes CPU to validate.

So that leaves us with the choice between caching negative results or caching only valid JWTs.

I've made cache implementation polymorphic over value computation monad so it is easy to change strategies. The latest commits introduce possibility of selecting one of two variants in PostgREST.Auth.JwtCache - it the matter of changing `cachingErrors` to `notCachingErrors` in `newJwtCache`. For now I've left both variants and disabled _unused local bindings_ warning to quiet linter. Once we decide which one is better we can remove the other one (or we can decide to make it configurable and leave both).

mkleczek commented:
> > So that leaves us with the choice between caching negative results or caching only valid JWTs.
> 
> I wonder whether we can cache all valid JWTs, but including expired ones? Aka when validation fails because of expiry, still cache. When validation fails because of something else, don't.

That’s exactly option 2. Which is currently implemented as notCachingErrors variant (switchable in JwtCache.init).

As described in the description of the PR (first comment) - we don’t cache AuthResults but raw parsed claims that are re-validated for each request.

mkleczek commented:
> > That’s exactly option 2.
> 
> Well, not exactly, but close enough I agree. We'd still cache those that fail, for example, and audience check or so. But that's totally fine, yes.

Right - it is not exactly the same.

The reason I decided to leave claims checking until after cache lookup are two-fold:

1. Time sensitive claims validation is dependent on well... time :) Not only `exp` but also for example `nbf` - it might become valid in the future.
2. On the other hand `aud` validation is configuration sensitive and I wanted to minimize the number of configuration options which require whole cache reset when changed.

mkleczek commented:
> > I wanted to minimize the number of configuration options which require whole cache reset when changed.
> 
> Yes, this makese a lot of sense!
> 
> So the only change that requires a reset is the change of secret, right?

That, and - of course - turning off caching alltogether (ie. setting jwt-cache-max-size=0)

mkleczek commented:
@steve-chavez @wolfgangwalther @taimoorzaeem 

After some more work on this PR I think it is now in a mergeable state (pending documentation changes, changelog adjustments etc. - and of course code review).

I am pretty confident it is working fine as I've added JWT cache behavior tests that verify hits/misses and evictions using metrics.

Please, let me know if there are any adjustments / changes required (or if you think the whole idea is wrong).

taimoorzaeem commented on src/PostgREST/Config.hs:
```diff
@@ -287,7 +289,8 @@ parser optPath env dbSettings roleSettings roleIsolationLvl =
           (optBool "jwt-secret-is-base64")
           (optBool "secret-is-base64"))
-    <*> (fromMaybe 0 <$> optInt "jwt-cache-max-lifetime")
+--    <*> (fromMaybe 0 <$> optInt "jwt-cache-max-lifetime")
+    <*> (fromMaybe 0 <$> optInt "jwt-cache-max-size")
```
So the `size` in there seems bit confusing as it may be confused with size in memory. How about renaming the config to `jwt-cache-max-entries`? It seems more appropriate, no?

mkleczek replied:
> So the `size` in there seems bit confusing as it may be confused with size in memory. How about renaming the config to `jwt-cache-max-entries`? It seems more appropriate, no?

Fine for me. @steve-chavez WDYT?

steve-chavez replied:
Yes, agree with `jwt-cache-max-entries`.

mkleczek replied:
Changed to `jwt-cache-max-entries`

steve-chavez commented on test/spec/Feature/Auth/JwtCacheSpec.hs:
```diff
@@ -0,0 +1,167 @@
+spec :: SpecWith (MetricsState, Application)
+spec = describe "Server started with JWT and metrics enabled" $ do
+  it "Should not have JWT in cache" $ do
+    let auth = genToken [json|{"exp": 9999999999, "role": "postgrest_test_author", "id": "jdoe1"}|]
+
+    expectCounters
+      [
+        requests (+ 1)
+      , hits     (+ 0)
+      ] $
```
This is a really elegant way to test the metrics :100: So easy to read!

steve-chavez commented on test/spec/Feature/Auth/JwtCacheSpec.hs:
```diff
@@ -0,0 +1,167 @@
+         request methodGet "/authors_only" [auth] "" `shouldRespondWith` 401
+      *> request methodGet "/authors_only" [auth] "" `shouldRespondWith` 401
+
+  it "Should evict entries from the JWT cache" $ do
```
Maybe inline the conf for easier readability.
 
```suggestion
  it "Should evict entries from the JWT cache (jwt cache max is 2)" $ do
```

steve-chavez commented on src/PostgREST/CLI.hs:
```diff
@@ -203,8 +203,8 @@ exampleConfigFile =
       |
-      |## Enables and set JWT Cache max lifetime, disables caching with 0
-      |# jwt-cache-max-lifetime = 0
+      |## Enables JWT Cache and sets its max size, disables caching with 0
+      |# jwt-cache-max-size = 0
```
I think we should make the default 1000. That means the cache will be enabled by default for next major.

steve-chavez replied:
1000 is a rough estimation, mentioned before on https://github.com/PostgREST/postgrest/pull/3802#issuecomment-2755566175

mkleczek replied:
> I think we should make the default 1000. That means the cache will be enabled by default for next major.

JWT cache makes sense **only** with asymmetric keys so I wouldn't turn it on by default

mkleczek replied:
> I think we should make the default 1000. That means the cache will be enabled by default for next major.

https://github.com/PostgREST/postgrest/pull/4084#issuecomment-2915678736

mkleczek commented:
> @mkleczek I've noticed that the [jwt loadtest results](https://github.com/PostgREST/postgrest/actions/runs/15274145867?pr=4084) show a perf drop compared to main and latest version. Is that expected?

Our load test is the worst possible case for this (and I would say any) cache: all JWTs are different so no caching but - if the cache is smaller than the number of requests then we pay the price of eviction and garbage collection.

mkleczek commented:
@steve-chavez @wolfgangwalther 

Looks like combining the default cache size from https://github.com/PostgREST/postgrest/pull/4084#discussion_r2110593171 with auto-configuration of the cache idea from https://github.com/PostgREST/postgrest/pull/4084#issuecomment-2915276059 allows us to have a pretty good user experience:

1. If max size is **not** configured, perform auto-configuration that turns on JWT cache of size 1000 but **only** if we have asymmetric keys in configuration.
2. Max size of <= 0 disables caching regardless of the configured keys.
3. Max size > 0 forces caching regardless of the configured keys.

Implemented in https://github.com/PostgREST/postgrest/pull/4084/commits/c6e73b13f41c19750aff56e13d56f265d0b68ca7

steve-chavez commented:
> updating yet another metric (and of questionable value) would eat up additional cycles

Cool. Fair enough.

> This is somewhat surprising - we do not remove from the cache explicitly right now, so it should be always full (hence occupying constant memory). Maybe what you see is GC effects?

Right. It looks like it's indeed GC as RSS and VSZ both decrease simultaneously as seen on:

[Screencast from 05-28-2025 02:12:05 PM.webm](https://github.com/user-attachments/assets/434f1edb-2341-4450-8a9a-12c58408fbbd)

---

I detected a problem with the previous JWT cache that this PR solves https://github.com/PostgREST/postgrest/issues/4107.

mkleczek commented:
> > Our load test is the worst possible case for this (and I would say any bounded) cache: all JWTs are different so no caching
> 
> With the new default, we should change the JWT load test to use asymmetric keys - and then see how much perf we lose in that worst case.

@wolfgangwalther @steve-chavez  @taimoorzaeem
See https://github.com/PostgREST/postgrest/pull/4084/commits/4d7ccc431409bc8c35335e1609ab3b4b63c23b8a

I've changed load test to use 1000 JWTs signed by RSA4096 key. Also added additional parameter `--jwtcache=off` to turn off JWT caching to compare results.

In case of asymmetric RSA caching gives around 4x performance boost (all response times up to 95th percentile are 4x lower with caching turned on).

steve-chavez commented:
> I've changed load test to use 1000 JWTs signed by RSA4096 key. Also added additional parameter --jwtcache=off to turn off JWT caching to compare results.

@mkleczek Cool! Can you split those commits to another PR for easier review/merge?

mkleczek commented on src/PostgREST/Metrics.hs:
```diff
@@ -19,22 +20,36 @@ import PostgREST.Observation
+    jwtCacheHits         :: Counter,
+    jwtCacheEvictions    :: Counter
+  }
 
 init :: Int -> IO MetricsState
 init configDbPoolSize = do
-  poolTimeouts <- register $ counter (Info "pgrst_db_pool_timeouts_total" "The total number of pool connection timeouts")
-  poolAvailable <- register $ gauge (Info "pgrst_db_pool_available" "Available connections in the pool")
-  poolWaiting <- register $ gauge (Info "pgrst_db_pool_waiting" "Requests waiting to acquire a pool connection")
-  poolMaxSize <- register $ gauge (Info "pgrst_db_pool_max" "Max pool connections")
-  schemaCacheLoads <- register $ vector "status" $ counter (Info "pgrst_schema_cache_loads_total" "The total number of times the schema cache was loaded")
-  schemaCacheQueryTime <- register $ gauge (Info "pgrst_schema_cache_query_time_seconds" "The query time in seconds of the last schema cache load")
-  setGauge poolMaxSize (fromIntegral configDbPoolSize)
-  pure $ MetricsState poolTimeouts poolAvailable poolWaiting poolMaxSize schemaCacheLoads schemaCacheQueryTime
+  metricState <- MetricsState <$>
+    register (counter (Info "pgrst_db_pool_timeouts_total" "The total number of pool connection timeouts")) <*>
+    register (gauge (Info "pgrst_db_pool_available" "Available connections in the pool")) <*>
+    register (gauge (Info "pgrst_db_pool_waiting" "Requests waiting to acquire a pool connection")) <*>
+    register (gauge (Info "pgrst_db_pool_max" "Max pool connections")) <*>
+    register (vector "status" $ counter (Info "pgrst_schema_cache_loads_total" "The total number of times the schema cache was loaded")) <*>
+    register (gauge (Info "pgrst_schema_cache_query_time_seconds" "The query time in seconds of the last schema cache load")) <*>
+    register (counter (Info "pgrst_jwt_cache_requests_total" "The total number of JWT cache lookups")) <*>
+    register (counter (Info "pgrst_jwt_cache_hits_total" "The total number of JWT cache hits")) <*>
+    register (counter (Info "pgrst_jwt_cache_evictions_total" "The total number of JWT cache evictions"))
+  setGauge (poolMaxSize metricState) (fromIntegral configDbPoolSize)
+  pure metricState
```
https://github.com/PostgREST/postgrest/pull/4120

taimoorzaeem commented on src/PostgREST/Config.hs:
```diff
@@ -97,7 +97,7 @@ data AppConfig = AppConfig
   , configJwtSecret                :: Maybe BS.ByteString
   , configJwtSecretIsBase64        :: Bool
-  , configJwtCacheMaxLifetime      :: Int
+  , configJwtCacheMaxSize          :: Maybe Int
```
Not that it matters much, but for the sake of consistency:
```suggestion
  , configJwtCacheMaxEntries       :: Maybe Int
```

mkleczek replied:
Good catch, fixed.

taimoorzaeem commented on src/PostgREST/Observation.hs:
```diff
@@ -151,6 +153,10 @@ observationMessage = \case
     "Trying to borrow a connection from pool"
   PoolRequestFullfilled ->
     "Borrowed a connection from the pool"
+  JwtCacheLookup _ ->
+    "Looked up a JWT in JWT cache"
+  JwtCacheEviction ->
+    "Evicted entry from JWT cache"
```
I think we should only log these on `logLevel >= LogDebug`. You can add this check here:

https://github.com/PostgREST/postgrest/blob/f85113669d29962699f11cacb06f1953bac23254/src/PostgREST/Logger.hs#L100-L103

mkleczek replied:
Fixed.

mkleczek commented on src/PostgREST/Config.hs:
I thought about it but such separation gets messy when handling configuration update. You have to handle both `configJWKS` and `configJwtCacheMaxEntries` in `JwtCache` module anyway. Having this logic **only** there seems better encapsulated (but the price is - as you say - lack of default value in config dump).

Additionally - it would make configuration reading monadic (instead of applicative as it is right now). The question really is: how much logic should be in Config module - should it be "dumb" and only responsible for parsing config values and providing defaults while leaving interpretation and handling of dependencies between configuration properties to other modules? Or should it be "smart" and actually become an interpreter of some configuration language assembling application components based on it?

steve-chavez commented:
@mkleczek Just merged https://github.com/PostgREST/postgrest/pull/4142. Once you rebase the loadtest will now include a markdown table showing process monitoring (MEM and CPU) for each elapsed second, comparing head/main/latest-version.

This should prove that your PR now makes the JWT cache memory bound. I've tried this locally and it should show the MEM % staying at 0 % instead of always growing :raised_hands:

steve-chavez commented:
@mkleczek Actually I don't know if this is because the rebase is somehow wrong, note that it contains other commits (which in turn makes the PR contain unrelated changes - harder to review).

Could you correct the rebase so it only shows the relevant commits?

steve-chavez commented on nix/tools/loadtest.nix:
```diff
@@ -61,7 +61,6 @@ let
         export PGRST_DB_TX_END="rollback-allow-override"
         export PGRST_LOG_LEVEL="crit"
         export PGRST_JWT_SECRET="reallyreallyreallyreallyverysafe"
-        export PGRST_JWT_CACHE_MAX_LIFETIME="86400"
```
Ahhhh, nice catch. This is what caused the confusion on https://github.com/PostgREST/postgrest/issues/4151.

steve-chavez commented:
> What's more: in case of symmetric JWT keys I don't think cache lookup is faster than simply performing JWT verification.

Maybe we should actually confirm if the above is true? It would be less confusing for users to just have one default for both type of keys.

Now the jwt-hs loadtest results (worst case) is proving that the new JWT cache is more efficient than the old one ([results](https://github.com/PostgREST/postgrest/actions/runs/15803495465?pr=4084)).

Additionally on #4151, we were comparing new cache (head) vs old cache disabled (main/v13.0.4) and it showed virtually no difference in throughput or MEM/CPU consumption  ([results](https://github.com/PostgREST/postgrest/actions/runs/15781551670?pr=4153)).

So far it looks we should just enable for both symmetric and asymmetric for simplicity? 

---

> Maybe we should actually confirm if the above is true?

We could add one more loadtest for jwt-hs (not a worst case like it is now) to confirm.

mkleczek commented:
> > > What's more: in case of symmetric JWT keys I don't think cache lookup is faster than simply performing JWT verification.
> > > Maybe we should actually confirm if the above is true?
> > 
> > 
> > We could add one more loadtest for jwt-hs (not a worst case like it is now) to confirm.
> 
> There is `--jwtcache=off` switch available in `jwt-hs` and `jwt-rsa` load tests but I am not sure how to add additional test run.

@steve-chavez 
So I've run some experiments locally with 1000 symmetric JWTs and compared caching and non-caching configuration. I found no difference in response times (even though all JWTs were cached). In other words caching did not help but also did not add any noticeable overhead.

I would say it is your call if we should get rid of auto-configuring cache for asymmetric keys and leaving it off for symmetric. (Although I personally like the idea of "smart" defaults and "smart" auto configuration in general - just let the software "do the right thing").
If you think it is more confusing to be "smart" then let's be it.

WDYT?

steve-chavez commented:
> So I've run some experiments locally with 1000 symmetric JWTs and compared caching and non-caching configuration. I found no difference in response times (even though all JWTs were cached). In other words caching did not help but also did not add any noticeable overhead.

@mkleczek Do you think the JWT cache could improve perf for symmetric JWTs in the future? Maybe it can be improved later somehow?

> If we're certain of that, then there is no reason to be able to disable the cache at all, right? Would it make the code simpler and thus easier to maintain?

Right, I think maintainability it's a stronger reason to not have smart defaults.

mkleczek commented:
@wolfgangwalther @steve-chavez 

As requested I removed "smart" auto configuration of JWT cache.
Right now JWT cache of size 1000 is enabled by default regardless of JWT key. It can be turned off by setting `jwt-cache-max-entries = 0`

mkleczek commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   + It now shows the invalid schema in the `message` field.
   + The exposed schemas are now listed in the `hint` instead of the `message` field.
 - Improve error details of `PGRST301` error by @taimoorzaeem in #4051
+- JWT cache based on sieve algorithm
```
Fixed

steve-chavez commented:
> @steve-chavez
So I've run some experiments locally with 1000 symmetric JWTs and compared caching and non-caching configuration. I found no difference in response times (even though all JWTs were cached). In other words caching did not help but also did not add any noticeable overhead.

I've finished the jwt loadtests and now the above can be proven on CI too: https://github.com/PostgREST/postgrest/pull/4175

@mkleczek Could you rebase here and correct the failing io test?

> I also updated documentation but bear with me, I'm not a native speaker and it might need some work.

I believe no one here is a native speaker :smiley:. Code LGTM so next I'll review the docs and then we can merge.

steve-chavez commented:
Just managed to understand the failure on CI, it looks like the JWT cache is not completely disabled when having `PGRST_JWT_CACHE_MAX_ENTRIES: 0`.

To reproduce:

```bash
# nix-shell
PGRST_LOG_LEVEL=debug PGRST_JWT_CACHE_MAX_ENTRIES=0 postgrest-with-postgresql-16 -f test/spec/fixtures/load.sql postgrest-run 

curl localhost:3000/authors_only -H "Authorization: Bearer $(postgrest-gen-jwt postgrest_test_author)"

16/Jul/2025:18:10:45 -0500: Schema cache loaded 271 Relations, 232 Relationships, 148 Functions, 15 Domain Representations, 45 Media Type Handlers, 1196 Timezones
16/Jul/2025:18:10:45 -0500: Schema cache loaded in 15.8 milliseconds
16/Jul/2025:18:10:46 -0500: Looked up a JWT in JWT cache
```

steve-chavez commented on src/PostgREST/Auth/JwtCache.hs:
```diff
@@ -1,99 +1,112 @@
+  cached = SC.cached
+
+decode :: JwtCache -> ByteString -> ExceptT Error IO JSON.Object
+decode JwtNoJwks        = const $ throwError (JwtErr JwtSecretMissing)
```
The `JwtSecretMissing` doesn't seem right on the `JwtCache` module, it looks to me it should be handled on a previous step. 

I just realized some commented out code was left on main: https://github.com/PostgREST/postgrest/blob/2d4967c4b40faea70a1c0464eac0cd85c6e5a88b/src/PostgREST/Auth/Jwt.hs#L102-L106

Done on https://github.com/PostgREST/postgrest/commit/a409a2cb9487216a4dba6d955a79a6f1121d5edf

Which also moved the `JwtSecretMissing` from Jwt.hs, which looked like a better place for it.

mkleczek replied:
@steve-chavez
Sorry about delays in taking care of it - I am on holidays until next week. Will take care of the issues once I'm back.

steve-chavez commented on src/PostgREST/Auth/JwtCache.hs:
```diff
@@ -1,99 +1,112 @@
+            -- select cachingErrors or notCachingErrors
+            notCachingErrors (readTVar maxSizeTVar) key
+
+    cachingErrors :: STM Int -> JwkSet -> IO (SC.Cache IO ByteString (Either Error JSON.Object))
+    cachingErrors maxSize key = SC.cacheIO (SC.CacheConfig maxSize
+            (runExceptT . parseAndDecodeClaims key)
+            (observationHandler . JwtCacheLookup) -- lookup metrics
+            (const . const $ observationHandler JwtCacheEviction) -- evictions metrics
+            alwaysValid) -- no invalidation for now
```
We should now delete this and remove the top `{-# OPTIONS_GHC -Wno-unused-local-binds #-}` right?

steve-chavez commented on src/PostgREST/Cache/Sieve.hs:
```diff
@@ -0,0 +1,218 @@
+cache cacheConfig = mdo
+    tail <- newTVar (Some head)
+    entries <- SH.new
+    finger <- newTVar tail
```
slightly pedantic, this should be `hand` right?  Just understood the algorithm from the blog post and they use the term `hand`.

steve-chavez commented on test/io/test_io.py:
```diff
@@ -1838,10 +1838,10 @@ def test_log_pool_req_observation(level, defaultenv):
-            output = postgrest.read_stdout(nlines=4)
-            assert pool_req in output[0]
-            assert pool_req_fullfill in output[3]
-            assert len(output) == 4
+            output = postgrest.read_stdout(nlines=5)
+            assert pool_req in output[1]
+            assert pool_req_fullfill in output[4]
+            assert len(output) == 5
```
@mkleczek Just confirmed the problem on https://github.com/PostgREST/postgrest/pull/4084#issuecomment-3081692509 doesn't happen anymore :tada: 

So why do we need to adjust the traces here? I'd suggest just disabling the cache for the `test_log_pool_req_observation` test.

mkleczek replied:
I've added debug logging for JWT cache lookup so the test failed as there was one additional line.

steve-chavez commented on docs/references/auth.rst:
```diff
@@ -98,9 +98,11 @@ The ``Bearer`` header value can be used with or without capitalization(``bearer`
 
-PostgREST validates ``JWTs`` on every request. We can cache ``JWTs`` to avoid this performance overhead.
+PostgREST validates ``JWTs`` on every request. Signature validation (especially asymmetric such as RSA) is slow and we can cache ``JWT`` validation results to avoid this performance overhead.
 
-To enable JWT caching, the config :code:`jwt-cache-max-lifetime` is to be set. It is the maximum number of seconds for which the cache stores the JWT validation results. The cache uses the :code:`exp` claim to set the cache entry lifetime. If the JWT does not have an :code:`exp` claim, it uses the config value. See :ref:`jwt-cache-max-lifetime` for more details.
+JWT caching is automatically enabled unless the config :code:`jwt-cache-max-entries` is set to 0.
+
+See :ref:`jwt-cache-max-entries` for more details.
```
I think we're missing some details in the docs, I'll help and make a PR to your repo to see if you agree.

mkleczek replied:
That would be great, thanks a lot!

steve-chavez replied:
Added more details on https://postgrest--4228.org.readthedocs.build/en/4228/references/auth.html#jwt-cache

Had to rebase since I shuffled some sections in the auth page, see: https://github.com/PostgREST/postgrest/pull/4228

LMK if it looks good.

mkleczek replied:
@steve-chavez 
I've merged `steve-chavez:jwt-cache-sieve-docs` here.

mkleczek replied:
> @mkleczek Made a PR on [mkleczek#1](https://github.com/mkleczek/postgrest/pull/1)

Merged

steve-chavez reviewed (approved):
Awesome work! :100:

steve-chavez commented:
:raised_hands: We're now included in the "Adoption" section of https://cachemon.github.io/SIEVE-website/ :raised_hands:
